cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, List, Optional, Sequence, Union
+from scipy.linalg import expm
+import numpy as np
+
+from cirq import devices, ops, protocols, qis
+from cirq.devices.noise_utils import (
+    PHYSICAL_GATE_TAG,
+)
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _left_mul(mat: np.ndarray) -> np.ndarray:
+    """Superoperator associated with left multiplication by a square matrix."""
+    mat = np.asarray(mat)
+    if mat.shape[-1] != mat.shape[-2]:
+        raise ValueError(
+            f'_left_mul only accepts square matrices, but input matrix has shape {mat.shape}.'
+        )
+    dim = mat.shape[-1]
+
+    return np.kron(mat, np.eye(dim))
+
+
+def _right_mul(mat: np.ndarray) -> np.ndarray:
+    """Superoperator associated with right multiplication by a square matrix."""
+    mat = np.asarray(mat)
+    if mat.shape[-1] != mat.shape[-2]:
+        raise ValueError(
+            f'_right_mul only accepts square matrices, but input matrix has shape {mat.shape}.'
+        )
+    dim = mat.shape[-1]
+
+    return np.kron(np.eye(dim), np.swapaxes(mat, -2, -1))
+
+
+def _lindbladian(left_op: np.ndarray) -> np.ndarray:
+    r"""Superoperator representing a Lindbladian.
+
+    The Lindbladian generated by a single operator A is the superoperator
+
+        $$
+        L(\rho) = A \rho A^\dagger - 0.5 (A^\dagger A \rho + \rho A^\dagger A)
+        $$
+
+    Args:
+        left_op: The operator acting on the left in the Lindbladian (A above).
+
+    Returns:
+        Superoperator corresponding to the Lindbladian.
+    """
+    left_op = np.asarray(left_op)
+    right_op = left_op.conj().T
+    square = right_op @ left_op
+    out = _left_mul(left_op) @ _right_mul(right_op)
+    out -= 0.5 * (_left_mul(square) + _right_mul(square))
+    return out
```

Comment:
Looking at _left_mul and _right_mul and the action they have on the matrices. You can cut out this: `out = _left_mul(left_op) @ _right_mul(right_op)` in favor of just doing a scalar multiply:
```
In [4]: %%timeit
   ...: 
   ...: np.einsum('i,jk->ijk', u.reshape(-1), u)
2.87 µs ± 16.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %%timeit
   ...: np.kron(u, np.eye(2)) @ np.kron(np.eye(2), u)
   ...: 
   ...: 
37.1 µs ± 241 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

```
A factor of 10 on something like this seems worth it while also being able to get rid of left and right mul.

Reply:
The `-= 0.5 * (_left_mul(square) + _right_mul(square))` might also be able to be replaced by a few in-place reshapes + additions instead of again kron'ing up.

Reply:
Discussed offline; I believe the decision there was to leave these as-is for now to make this as direct of a port from the internal tools as possible, and follow up with potential optimizations later. I've opened #4768 to track those.